File: solutions/ruby/blackjack/1/blackjack.rb
module Blackjack
  def self.parse_card(card)
    case card
    when "ace"
      11
    when "two"
      2
    when "three"
      3
    when "four"
      4
    when "five"
      5
    when "six"
      6
    when "seven"
      7
    when "eight"
      8
    when "nine"
      9
    when "ten"
      10
    when "jack", "queen", "king"
      10
    else
      0
    end
  end

  def self.card_range(card1, card2)
    card1_value = parse_card(card1)
    card2_value = parse_card(card2)

    case card1_value + card2_value
    when (4..11)
      "low"
    when (12..16)
      "mid"
    when (17..20)
      "high"
    else
      "blackjack"
    end
  end

  def self.first_turn(card1, card2, dealer_card)
    dealer_card_value = parse_card(dealer_card)
    card_range = card_range(card1, card2)

    case
    when (card1 == 'ace' && card1 == card2)
      "P"
    when (card_range == 'blackjack' && dealer_card_value < 10)
      "W"
    when (card_range == 'blackjack' && dealer_card_value >= 10) || (card_range == 'high') || (card_range == 'mid' && dealer_card_value < 7)
      "S"
    when (card_range == 'mid' && dealer_card_value >= 7) || (card_range == 'low')
      "H"
    else
      raise "Error no expected"
    end
  end
end

File: solutions/go/blackjack/1/blackjack.go
package blackjack

// ParseCard returns the integer value of a card following blackjack ruleset.
func ParseCard(card string) int {
	switch card {
	case "ace":
		return 11
	case "two":
		return 2
	case "three":
		return 3
	case "four":
		return 4
	case "five":
		return 5
	case "six":
		return 6
	case "seven":
		return 7
	case "eight":
		return 8
	case "nine":
		return 9
	case "ten", "jack", "queen", "king":
		return 10
	}

	return 0
}

// FirstTurn returns the decision for the first turn, given two cards of the
// player and one card of the dealer.
func FirstTurn(card1, card2, dealerCard string) string {

	sumUp := ParseCard(card1) + ParseCard(card2)
	switch {
	case card1 == "ace" && card2 == "ace":
		return "P"
	case sumUp == 21 && !(dealerCard == "ace" || dealerCard == "kind" || dealerCard == "queen" || dealerCard == "jack"):
		return "W"

	case sumUp >= 12 && sumUp <= 16 && ParseCard(dealerCard) >= 7:
		return "H"

	case sumUp <= 11:
		return "H"

	}

	return "S"
}
